Sources/Network/JSON.swift

```diff
@@ -304,13 +307,14 @@ public enum JSON {
     ///   - parseAPIError: The API error parsing closure.
     /// - Returns: The Date obtained from the formatter or nil.
     /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
-    public static func parseOptionalDateAttribute<T>(key: JSON.AttributeKey,
+    public static func parseOptionalDateAttribute<T>(_ key: JSON.AttributeKey,
                                                      json: JSON.Dictionary,
                                                      formatter: (T) -> Date?,
                                                      parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date? {
 
         guard let rawValue = json[key] else {
-            throw parseAPIError?(json) ?? Error.missingAttribute(key, json: json)
+            if let apiError = parseAPIError?(json) { throw apiError }
+            return nil
         }
 
         let jsonValue: T = try parseValue(rawValue: rawValue, key: key, json: json, parseAPIError: parseAPIError)
```

Comment:
missing the above check here on the `formatter()` call too 😉

Reply:
what do you mean? should we call the `apiError` when the formatter fails?

Reply:
From my perspective, the `parseOptional` should only return `nil` if the key doesn't come in the JSON at all, and not in any other case.

So, if the key is present in the JSON, we should treat it as a normal `parseAttribute`, and in this case if the `formatter()` call fails and returns `nil`, it means that the data was invalid and we should notify the caller. Makes sense? 😛